iml-agent/src/action_plugins/stratagem/action_filesync.rs

```diff
@@ -50,179 +71,112 @@ fn do_rsync<'a>(
-        let output;
-        if md.is_dir() || (md.len() > LARGEFILE) {
-            output = Command::new(&mpi_path)
-                .arg("--allow-run-as-root")
-                .arg("-c")
-                .arg(format!("{}", mpi_count))
-                .arg("--hostfile")
-                .arg("/etc/iml/filesync-hostfile")
-                .arg("dsync")
-                .arg("-S")
-                .arg(dest_file)
-                .arg(src_file)
-                .kill_on_drop(true)
-                .output();
-        } else {
-            output = Command::new("rsync")
-                .arg(dest_file)
-                .arg(src_file)
-                .kill_on_drop(true)
-                .output();
-        }
-
-        let output = output.await?;
+                output.status.code()
+            );
 
-        if !output.status.success() {
-            // TODO: handle termination conditions without defaulting to 0
-            let res = FidError {
-                fid: fid.fid.clone(),
-                data: fid.data.clone(),
+            Ok::<_, ImlAgentError>(FidError {
+                fid: work.fid.fid.clone(),
+                data: work.fid.data.clone(),
                 errno: output.status.code().unwrap_or(1) as i16,
-            };
-            result.push(res);
-        }
-        tracing::debug!(
-            "exited with {} {} {}",
-            output.status,
-            std::str::from_utf8(&output.stdout)?,
-            std::str::from_utf8(&output.stderr)?
-        );
-    }
+            })
+        })
+        .chunks(1)
+        .map(Ok)
+        .try_fold(vec![], |mut acc, xs| async {
+            let mut xs = try_join_all(xs).await?;
 
-    Ok(result)
-}
+            xs.retain(|x| x.errno != 0);
+            acc.extend(xs);
 
-arg_enum! {
-    #[derive(Debug, serde::Deserialize, Clone, Copy, PartialEq)]
-    #[serde(rename_all = "lowercase")]
-    pub enum ActionType {
-    Push,
-    Pull,
-    }
+            Ok::<_, ImlAgentError>(acc)
+        })
 }
 
 #[derive(Debug, serde::Deserialize)]
 pub struct TaskArgs {
     pub remote: String,
     pub action: ActionType,
 }
+
 /// Process FIDs
 pub async fn process_fids(
     (fsname_or_mntpath, task_args, fid_list): (String, TaskArgs, Vec<FidItem>),
 ) -> Result<Vec<FidError>, ImlAgentError> {
     let llapi = search_rootpath(fsname_or_mntpath).await?;
-    match task_args.action {
-        ActionType::Push => archive_fids(llapi.clone(), &task_args.remote, fid_list).await,
-        ActionType::Pull => restore_fids(llapi.clone(), &task_args.remote, fid_list).await,
+    let mut rsync_list: Vec<Work> = vec![];
+    let mut dsync_list: Vec<Work> = vec![];
+    let mut result = vec![];
+
+    let src_root: PathBuf;
+    let dest_root: PathBuf;
+    if task_args.action == ActionType::Push {
+        src_root = llapi.mntpt().into();
+        dest_root = task_args.remote.into();
+    } else {
+        src_root = task_args.remote.into();
+        dest_root = llapi.mntpt().into();
     }
```

Comment:
```suggestion
    let (src_root, dest_root): (PathBuf, PathBuf) = if task_args.action == ActionType::Push {
        (llapi.mntpt().into(), task_args.remote.into())
    } else {
        (task_args.remote.into(), llapi.mntpt().into())
    };
```

Is a bit more idiomatic than uninitialized bindings

Reply:
Done